kernel/src/vm/vmar/mod.rs

```diff
@@ -364,59 +364,45 @@ impl Vmar_ {
         Ok(())
     }
 
-    fn check_overwrite(&self, mapping_range: Range<usize>, can_overwrite: bool) -> Result<()> {
-        let inner = self.inner.read();
-
-        if !can_overwrite && inner.vm_mappings.find(&mapping_range).next().is_some() {
-            return_errno_with_message!(
-                Errno::EACCES,
-                "mapping range overlapped with another mapping"
-            );
-        }
-
-        Ok(())
-    }
-
     /// Returns the attached `VmSpace`.
     fn vm_space(&self) -> &Arc<VmSpace> {
         &self.vm_space
     }
 
-    /// Maps a `VmMapping` to this VMAR.
-    fn add_mapping(&self, mapping: VmMapping) {
-        self.inner.write().vm_mappings.insert(mapping);
-    }
-
-    fn allocate_free_region_for_mapping(
+    /// Build the `VmMapping`.
+    ///
+    /// Allocates a free region for the mapping, sets the map address,
+    /// and adds it to the VMAR.
+    fn build_mapping(
```

Comment:
My previous intention was to change the signature of this API to `pub fn build_mapping<R1, R2>(option: VmarMapOptions<R1, R2>)` (or only let it public for `vm_mapping` module). This way, the semantics of `build_mapping` within a `Vmar` would be more natural, as it is natural to manage a mapping within a Vmar when building one. This API would only need to receive a `VmarMapOptions` as a parameter, just like `VmMapping::build_mapping` did before.

This would help avoid the use of `set_map_to_addr`, which I also think somewhat peculiar.

Reply:
Moving the function of building `VmMapping` based on `VmarMapOptions` to `Vmar` requires making all fields public to `Vmar`, and modifying `VmarMapOptions::parent` to `Option<>`, which may look a bit ugly.